---
title: "Section 2.3 Results"
output:
  html_document:
    df_print: paged
---

# The Chinese hamster genome harbours thousands of short open reading frames

## Prepare for analysis
### Load libraries & functions
```{r include=FALSE}
package_list <- c(
  "tidyverse", "ggpubr", "ggpp", "Biostrings", "viridis"
)

lapply(package_list, require, character.only = TRUE)
```

### Create results directory
```{r}
results_dir <- "../section2.3/"
if (!dir.exists(results_dir)) {
  dir.create(results_dir, recursive = TRUE)
}
```


### Import the ORFRATER results
```{r}
load("../section2.2/results_2_2.RData")
```

###  ORF length plot
```{r}
source(file = "../../scripts/facetzoom2.R")
safe_colorblind_palette <- c(
  "#88CCEE", "#CC6677", "#888888", "#117733", "#332288", "#035968",
  "#AA4499", "#999933", "#882255", "#661100", "#6699CC", "#888888"
)

orf_length_plot <- table_s3 %>%
  mutate(`ORF type` = fct_relevel(`ORF type`,
    levels = "Extension", "Annotated", "Isoform",
    "Stop overlap", "Start overlap", "New", "Downstream", "Upstream"
  )) %>%
  ggplot(aes(y = `Length (AAs)`, x = `ORF type`, fill = `ORF type`)) +
  geom_boxplot(position = "dodge", size = 0.2, outlier.size = 0.5) +
  geom_jitter(size = 0.1, alpha = 0.1) +
  theme_bw() +
  scale_fill_manual(values = safe_colorblind_palette) +
  theme(axis.text.x = element_text(angle = 60, hjust = 1)) +
  facet_zoom2(ylim = c(0, 100)) +
  labs(x = "")

orf_length_plot
ggsave(plot = orf_length_plot, filename = paste(results_dir, "Figure 3A.png", sep = ""), width = 6, height = 4, device = "png", dpi = 700)
```


## Length distribution plots
```{r}
uorf_length_plot <- table_s3 %>%
  filter(`ORF type` == "Upstream") %>%
  ggplot(aes(x = sort(`Length (AAs)`))) +
  geom_histogram(bins = 50, fill = "#999933") +
  labs(y = "# uORFs", x = "uORF Length (aa)") +
  geom_text_npc(aes(npcx = 0.9, npcy = 0.9, label = paste0("Mean length:\n", round(mean(`Length (AAs)`)), " aa")), size = 3) +
  theme_bw()

ouorf_length_plot <- table_s3 %>%
  filter(`ORF type` == "Start overlap") %>%
  ggplot(aes(x = sort(`Length (AAs)`))) +
  geom_histogram(bins = 50, fill = "#332288") +
  labs(y = "# ouORFs", x = "ouORF Length (aa)") +
  theme_bw() +
  geom_text_npc(aes(npcx = 0.9, npcy = 0.9, label = paste0("Mean length:\n", round(mean(`Length (AAs)`)), " aa")), size = 3)

new_orf_length_plot <- table_s3 %>%
  filter(`ORF type` == "New") %>%
  ggplot(aes(x = sort(`Length (AAs)`))) +
  geom_histogram(bins = 50, fill = "#35968CFF") +
  labs(y = "# New ORFs", x = "ORF Length (aa)") +
  geom_text_npc(aes(npcx = 0.9, npcy = 0.9, label = paste0("Mean length:\n", round(mean(`Length (AAs)`)), " aa")), size = 3) +
  theme_bw()

figure_s3 <- ggarrange(uorf_length_plot,
  ggplot() +
    theme_void(),
  ouorf_length_plot, ggplot() +
    theme_void(),
  new_orf_length_plot,
  ncol = 5, widths = c(1, 0.05, 1, 0.05, 1)
)

figure_s3
ggsave(plot = figure_s3, filename = paste(results_dir, "Figure S3.png", sep = ""), dpi = 700, device = "png", width = 9, height = 3)
```

### sORF proportion
```{r}
all_short_orfs <- table_s3 %>%
  filter(`Length (AAs)` < 100)

table(all_short_orfs$`ORF type`) / sum(table(all_short_orfs$`ORF type`))


table(all_short_orfs$`ORF type`) / table(table_s3$`ORF type`)

table_s3 %>%
  filter(`Length (AAs)` < 100) %>%
  summarise(median(`Length (AAs)`))


sORF_codon <- table_s3 %>%
  mutate(length_classification = case_when(
    `Length (AAs)` >= 100 ~ ">= 100aa",
    TRUE ~ "< 100aa"
  )) %>%
  mutate(length_classification = fct_rev(as.factor(length_classification))) %>%
  filter(`ORF type` == "Upstream" | `ORF type` == "Start overlap" | `ORF type` == "New")

table(sORF_codon$`ORF type`,sORF_codon$`Start codon`)

proportion_sorf <- table_s3 %>%
  mutate(length_classification = case_when(
    `Length (AAs)` >= 100 ~ ">= 100aa",
    TRUE ~ "< 100aa"
  )) %>%
  mutate(length_classification = fct_rev(as.factor(length_classification))) %>%
  filter(`ORF type` == "Upstream" | `ORF type` == "Start overlap" | `ORF type` == "New") %>%
  ggplot(aes(x = `ORF type`, fill = length_classification)) +
  geom_bar(position = "fill") +
  scale_fill_manual(values = c("#FDE725FF", "#482677FF")) +
  theme_bw() +
  labs(x = "", y = "Proportion of ORFs", fill = "") +
  theme(axis.text.x = element_text(angle = 60, hjust = 1))
proportion_sorf
ggsave(plot = proportion_sorf, filename = paste(results_dir, "Figure 3B.png", sep = ""), width = 3, height = 3.5, device = "png", dpi = 700)
```
### Amino acid frequencies
```{r}
fasta <- readAAStringSet("../../amino_acid_analysis/long_orfs_for_AA_freq.fasta")
long_ORF <- colSums(alphabetFrequency(fasta)[, AA_STANDARD]) / (sum(colSums(alphabetFrequency(fasta)[, AA_STANDARD])))

fasta <- readAAStringSet("../../amino_acid_analysis/upstream_short_orfs_for_AA_freq.fasta")
short_upstream_ORF <- colSums(alphabetFrequency(fasta)[, AA_STANDARD]) / (sum(colSums(alphabetFrequency(fasta)[, AA_STANDARD])))

fasta <- readAAStringSet("../../amino_acid_analysis/non_coding_orfs.fasta")
short_ncRNA_ORF <- colSums(alphabetFrequency(fasta)[, AA_STANDARD]) / (sum(colSums(alphabetFrequency(fasta)[, AA_STANDARD])))

# display all AAs
amino_acid_plot_full <- bind_rows(long_ORF, short_upstream_ORF, short_ncRNA_ORF) %>%
  mutate(type = c("Annotated >= 100aa", "uORFs < 100aa", "New < 100aa")) %>%
  pivot_longer(names_to = "AA", values_to = "Freq", cols = -contains("type")) %>%
  mutate(aa_name = AMINO_ACID_CODE[AA]) %>%
  ggplot(aes(x = aa_name, y = Freq, fill = type)) +
  geom_col(width = 0.6, position = position_dodge(width = 0.6)) +
  theme_bw() +
  labs(x = "", y = "Frequency", fill = "") +
  scale_fill_manual(values = c("#CC79A7", "#35968CFF", "#999933")) +
  theme(legend.position = "right")
amino_acid_plot_full
ggsave(plot = amino_acid_plot_full, filename = paste(results_dir, "Figure S4.png", sep = ""), width = 7, height = 3.5, device = "png", dpi = 700)

# display AAs that are different
amino_acid_plot_selected <- bind_rows(long_ORF, short_upstream_ORF, short_ncRNA_ORF) %>%
  mutate(type = c("Annotated >= 100aa", "uORFs < 100aa", "New < 100aa")) %>%
  pivot_longer(names_to = "AA", values_to = "Freq", cols = -contains("type")) %>%
  mutate(aa_name = AMINO_ACID_CODE[AA]) %>%
  filter(aa_name != "His" & aa_name != "Leu" & aa_name != "Ser" & aa_name != "Val" & aa_name != "Thr" & aa_name != "Phe") %>%
  ggplot(aes(x = aa_name, y = Freq, fill = type)) +
  geom_col(width = 0.6, position = position_dodge(width = 0.6)) +
  theme_bw() +
  labs(x = "", y = "Frequency", fill = "") +
  scale_fill_manual(values = c("#CC6677", "#035968", "#999933")) +
  theme(legend.position = "right")

amino_acid_plot_selected
ggsave(plot = amino_acid_plot_selected, filename = paste(results_dir, "Figure 3C.png", sep = ""), width = 7, height = 3.5, device = "png", dpi = 700)
```

## Coverage track plots
### Load bedgraph data
```{r include=FALSE}
harr_min_nd <- read_tsv("../../alignment_tracks/merged/harr-nd.bedgraph",
  col_names = F
) %>% mutate(ribotype = "harr-nd")

chx_bedgraph <- read_tsv("../../alignment_tracks/merged/chx.p.site.transcriptome.bedgraph",
  col_names = F
) %>% mutate(ribotype = "chx")


chx_full_bedgraph <- read_tsv("../../alignment_tracks/merged/chx.fullcov.transcriptome.bedgraph",
  col_names = F
) %>% mutate(ribotype = "chx_full")

bedgraph_data <- bind_rows(harr_min_nd, chx_bedgraph, chx_full_bedgraph)

range01 <- function(x) {
  (x - min(x)) / (max(x) - min(x))
}
```

### DDIT3 uORF
```{r}
selected_transcript <- "XM_027392226.1"

current_orf <- table_s3 %>%
  filter(`Transcript ID` == selected_transcript)
current_orf


chx_p_scaled <- bedgraph_data %>%
  filter(X1 == selected_transcript) %>%
  filter(ribotype == "chx") %>%
  # filter(X2 <= max(current_orf$`Transcript stop position`) & X2 >= 0) %>%
  mutate(scale_bpm = range01(X4))

chx_full_scaled <- bedgraph_data %>%
  filter(X1 == selected_transcript) %>%
  filter(ribotype == "chx_full") %>%
  # filter(X2 <= 1056 & X2 >= 0) %>%
  mutate(scale_bpm = range01(X4))

track_data <- bind_rows(chx_p_scaled, chx_full_scaled)

track_data <- track_data %>%
  mutate(frame = ifelse(X2 >= current_orf$`Transcript start position` & X2 <= current_orf$`Transcript stop position` & ribotype == "chx", (X2 - current_orf$`Transcript start position`) %% 3,
    ifelse(X2 >= 167 & X2 <= 617 & ribotype == "chx", (X2 - 167) %% 3, NA)
  ))


chx_plot <- track_data %>%
  filter(ribotype == "chx" | ribotype == "chx_full") %>%
  ggplot(aes(x = X2, y = scale_bpm, fill = frame, alpha = ribotype)) +
  annotate("rect",
    xmin = c(current_orf$`Transcript start position`, 167), xmax = c(current_orf$`Transcript stop position`, 617), ymin = 0, ymax = 1,
    alpha = .1, fill = c("#999933", "#CC6677")
  ) +
  geom_bar(stat = "identity", position = "identity") +
  scale_fill_viridis() +
  # geom_vline(xintercept =c(36,141,168,617),  size=0.2, linetype=11) +
  scale_alpha_discrete(range = c(1, 0.3)) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0)) +
  theme_bw() +
  geom_text_npc(aes(npcx = 0.95, npcy = 0.95), label = "CHX", size = 6) +
  labs(
    x = "Transcript coordinates (nts)",
    y = "Ribo-seq (Scaled BPM)"
  )
chx_plot$layers <- c(geom_bar(stat = "identity", position = "identity"), chx_plot$layers)
chx_plot + theme(legend.position = "none")

ggsave(filename = paste(results_dir, "Figure 3D.png", sep = ""), width = 6, height = 3, device = "png", dpi = 700)
```
### Rad51d ouORF
```{r}
selected_transcript <- "XM_027426482.2"
current_orf <- table_s3 %>%
  filter(`Transcript ID` == selected_transcript)
current_orf

chx_p_scaled <- bedgraph_data %>%
  filter(X1 == selected_transcript) %>%
  filter(ribotype == "chx") %>%
  # filter(X2 <= max(current_orf$`Transcript stop position`) & X2 >= 0) %>%
  mutate(scale_bpm = range01(X4))

chx_full_scaled <- bedgraph_data %>%
  filter(X1 == selected_transcript) %>%
  filter(ribotype == "chx_full") %>%
  # filter(X2 <= 1056 & X2 >= 0) %>%
  mutate(scale_bpm = range01(X4))

track_data <- bind_rows(chx_p_scaled, chx_full_scaled)

track_data <- track_data %>%
  mutate(frame = ifelse(X2 >= current_orf$`Transcript start position` & X2 <= current_orf$`Transcript stop position` & ribotype == "chx", (X2 - current_orf$`Transcript start position`) %% 3,
    ifelse(X2 >= 191 & X2 <= 1180 & ribotype == "chx", (X2 - 191) %% 3, NA)
  ))


chx_plot <- track_data %>%
  filter(ribotype == "chx" | ribotype == "chx_full") %>%
  ggplot(aes(x = X2, y = scale_bpm, fill = frame, alpha = ribotype)) +
  annotate("rect",
    xmin = c(current_orf$`Transcript start position`, 191), xmax = c(current_orf$`Transcript stop position`, 1180), ymin = 0, ymax = 1,
    alpha = .1, fill = c("#332288", "#CC6677")
  ) +
  geom_bar(stat = "identity", position = "identity") +
  scale_fill_viridis() +
  # geom_vline(xintercept =c(36,141,168,617),  size=0.2, linetype=11) +
  scale_alpha_discrete(range = c(1, 0.3)) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0)) +
  theme_bw() +
  geom_text_npc(aes(npcx = 0.95, npcy = 0.95), label = "CHX", size = 6) +
  labs(
    x = "Transcript coordinates (nts)",
    y = "Ribo-seq (Scaled BPM)"
  )
chx_plot$layers <- c(geom_bar(stat = "identity", position = "identity"), chx_plot$layers)
chx_plot + theme(legend.position = "none")

ggsave(filename = paste(results_dir, "Figure 3E.png", sep = ""), width = 6, height = 3, device = "png", dpi = 700)
```

### XR_004770827.1_96502054_66aa
```{r message=FALSE, warning=FALSE}
selected_transcript <- "XR_004770827.1_96502054_66aa"

current_orf <- table_s3 %>%
  filter(`ORF-RATER name` == selected_transcript) %>%
  filter(`ORF type` == "New")
current_orf

chx_p_scaled <- bedgraph_data %>%
  filter(X1 == current_orf$`Transcript ID`) %>%
  filter(ribotype == "chx") %>%
  # filter(X2 <= max(current_orf$`Transcript stop position`) & X2 >= 0) %>%
  mutate(scale_bpm = range01(X4))

chx_full_scaled <- bedgraph_data %>%
  filter(X1 == current_orf$`Transcript ID`) %>%
  filter(ribotype == "chx_full") %>%
  # filter(X2 <= 1056 & X2 >= 0) %>%
  mutate(scale_bpm = range01(X4))

track_data <- bind_rows(chx_p_scaled, chx_full_scaled) %>%
  filter(X2 >= 0 & X2 <= 1000)

track_data <- track_data %>%
  mutate(frame = ifelse(X2 >= current_orf$`Transcript start position`[1] & X2 <= current_orf$`Transcript stop position`[1] & ribotype == "chx", (X2 - current_orf$`Transcript start position`[1]) %% 3, NA))

chx_plot <- track_data %>%
  filter(ribotype == "chx" | ribotype == "chx_full") %>%
  ggplot(aes(x = X2, y = scale_bpm, fill = frame, alpha = ribotype)) +
  annotate("rect",
    xmin = c(current_orf$`Transcript start position`), xmax = c(current_orf$`Transcript stop position`), ymin = 0, ymax = 1,
    alpha = .1, fill = c("#35968CFF")
  ) +
  geom_bar(stat = "identity", position = "identity") +
  scale_fill_viridis() +
  # geom_vline(xintercept =c(36,141,168,617),  size=0.2, linetype=11) +
  scale_alpha_discrete(range = c(1, 0.3)) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0)) +
  theme_bw() +
  geom_text_npc(aes(npcx = 0.95, npcy = 0.95), label = "CHX", size = 6) +
  labs(
    x = "Transcript coordinates (nts)",
    y = "RiboSeq coverage (BPM)",
    # title = "LOC103161792",
    subtitle = ""
  )
chx_plot$layers <- c(geom_bar(stat = "identity", position = "identity"), chx_plot$layers)
chx_plot + theme(legend.position = "none")

ggsave(filename = paste(results_dir, "Figure 3F.png", sep = ""), width = 6, height = 3, device = "png", dpi = 700)
```
