---
title: "Section 3.4 Results"
output:
  html_document:
    df_print: paged
---
# Prepare for analysis
```{r message=FALSE, warning=FALSE, include=FALSE}
package_list <- c(
  "tidyverse", "DESeq2", "writexl", "viridis", "ggpp"
)

suppressMessages(lapply(package_list, require, character.only = TRUE))

results_dir <- "/mnt/HDD2/colin/ribosome_footprint_profiling/results/section2.5/"
if (!dir.exists(results_dir)) {
  dir.create(results_dir, recursive = TRUE)
}
```

# Import the ORFRATER results
```{r}
load("../section2.2/results_2_2.RData")
```

# Analysis of New ORFs
## Transcript types
```{r}
new_orf_classification <- table_s3 %>%
  filter(`ORF type` == "New") %>%
  mutate(rna_type = case_when(
    str_detect(`ORF-RATER name`, "XR|NR") ~ "ncRNA",
    !str_detect(`ORF-RATER name`, "XR|NR") ~ "mRNA"
  )) %>%
  mutate(orf_length = case_when(
    `Length (AAs)` >= 5 & `Length (AAs)` &
      `Length (AAs)` <= 100 &
      rna_type == "ncRNA" ~ "lncRNA smORF <= 100aa",
    rna_type == "mRNA" ~ "mRNA",
    `Length (AAs)` < 5 | `Length (AAs)` &
      `Length (AAs)` > 100 &
      rna_type == "ncRNA" ~ "lncRNA ORF > 100aa"
  )) %>%
  mutate(
    type = "New ORFs",
    orf_length = factor(orf_length, levels = c("mRNA", "lncRNA ORF > 100aa", "lncRNA smORF <= 100aa"))
  )

table(new_orf_classification$rna_type, new_orf_classification$orf_length)

new_orf_classification_plot <- new_orf_classification %>%
  ggplot(aes(x = type, fill = orf_length)) +
  geom_bar(aes(y = (..count..) / sum(..count..)), position = "stack", width = 1) +
  theme_minimal() +
  scale_fill_manual(values = c("#999999", "#CC79A7", "#35968CFF")) +
  # scale_x_continuous(expand=c(0,0)) +
  scale_y_continuous(expand = c(0, 0)) +
  labs(fill = "", x = "", y = "Proportion of ORFs")
new_orf_classification_plot
ggsave(plot = new_orf_classification_plot, filename = paste(results_dir, "Figure_6A.png", sep = ""), dpi = 700, device = "png", width = 3, height = 3)
```
## Number of small ORFs per non-coding RNA
```{r}
lncRNA_smorfs <- table_s3 %>%
  filter(`ORF type` == "New" &
    str_detect(`ORF-RATER name`, "XR|NR") &
    `Length (AAs)` >= 5 & `Length (AAs)` <= 100) %>%
  group_by(`Transcript ID`) %>%
  mutate(num_ORFs = n()) %>%
  ungroup()

# plot the distribution of uORFs per_transcript
new_ORFs_per_transcript_plot <- lncRNA_smorfs %>%
  dplyr::select(`Transcript ID`, num_ORFs) %>%
  ggplot(aes(x = num_ORFs)) +
  geom_bar(stat = "count", fill = "#35968CFF") +
  scale_x_discrete(limits = as.character(c(1:max(lncRNA_smorfs$num_ORFs)))) +
  theme_bw() +
  labs(x = "ORFs per transcript", y = "Number of transcripts") +
  geom_text(stat = "count", aes(label = ..count..), vjust = -0.1, color = "black", size = 3)
new_ORFs_per_transcript_plot

ggsave(plot = new_ORFs_per_transcript_plot, filename = paste(results_dir, "Figure_6B.png", sep = ""), dpi = 700, device = "png", width = 3, height = 3)
```

```{r}
# 1. length distribution
lncRNA_smorf_length_plot <- lncRNA_smorfs %>%
  ggplot(aes(x = sort(`Length (AAs)`))) +
  geom_histogram(bins = 50, fill = "#35968CFF") +
  labs(y = "# lncRNA smORFs", x = "ORF Length (aa)") +
  geom_text_npc(aes(npcx = 0.9, npcy = 0.9, label = paste0("Mean length:\n", round(mean(lncRNA_smorfs$`Length (AAs)`)), " aa")), size = 3) +
  theme_bw()

lncRNA_smorf_length_plot
ggsave(plot = lncRNA_smorf_length_plot, filename = paste(results_dir, "Figure_6C.png", sep = ""), dpi = 700, device = "png", width = 3, height = 3)
```

# Load the DE results
```{r}
load(file = "../section2.4/results.2.4.RData")
```


## Filter thresholds 
```{r}
fold_change_threshold <- 1.5
pval_threshold <- 0.05
base_mean_threshold <- 0
average_counts <- 20 # set during Results 3.3
```

## New ORFs within AA length range
```{r}
smorf_ids<-table_s3 %>% filter(`ORF type` == "New" &
            str_detect(`ORF-RATER name`, "XR|NR") &
            `Length (AAs)` >= 5 & `Length (AAs)` <= 100 ) %>%
  dplyr::select(`ORF-RATER name`)
```


```{r}
sig_res_rnaseq_smorf <- res_rnaseq %>%
  as_tibble(rownames = "geneid") %>%
  filter(str_detect(geneid, "NR|XR")) %>%
  filter(geneid %in% smorf_ids$`ORF-RATER name`) %>%
  dplyr::filter(abs(log2FoldChange) >= log2(fold_change_threshold) &
    padj < pval_threshold & baseMean >= base_mean_threshold)

paste0(dim(sig_res_rnaseq_smorf)[1], " DE from RNA-seq")
paste0(sum(sig_res_rnaseq_smorf$log2FoldChange > 0), " up from RNA-seq")
paste0(sum(sig_res_rnaseq_smorf$log2FoldChange < 0), " down from RNA-seq")


rnaseq_volcano_plot <- res_rnaseq %>%
  as_tibble(rownames = "geneid") %>%
  drop_na(padj) %>%
  filter(str_detect(geneid, "NR|XR")) %>%
  mutate(diff_rnaseq = case_when(
    log2FoldChange >= log2(fold_change_threshold) & padj < pval_threshold & baseMean >= base_mean_threshold & str_detect(geneid, "NR|XR") == "TRUE" ~ "Up",
    log2FoldChange <= -log2(fold_change_threshold) & padj < pval_threshold & baseMean >= base_mean_threshold & str_detect(geneid, "NR|XR") == "TRUE" ~ "Down",
    TRUE ~ "No change"
  )) %>%
  ggplot(aes(x = log2FoldChange, y = -log10(padj), color = diff_rnaseq)) +
  geom_point(size = 1) +
  labs(x = "Log2 fold change (TS/NTS)", title = "RNA-seq", color = "") +
  theme_bw() +
  xlim(-2.5, 2.5) +
  geom_vline(
    xintercept = c(log2(fold_change_threshold), -log2(fold_change_threshold)),
    linetype = c(11, 11), size = c(0.2, 0.2)
  ) +
  geom_hline(yintercept = -log10(0.05), linetype = 11, size = 0.2) +
  scale_color_manual(values = c("#D55E00", "#999999", "#56B4E9"))
```

## RPF DE smORF
```{r}
sig_res_riboseq_smorf <- res_riboseq %>%
  as_tibble(rownames = "geneid") %>%
  filter(str_detect(geneid, "NR|XR")) %>%
  filter(geneid %in% smorf_ids$`ORF-RATER name`) %>%
  dplyr::filter(abs(log2FoldChange) >= log2(fold_change_threshold) &
    padj < pval_threshold & baseMean >= base_mean_threshold)

paste0(dim(sig_res_riboseq_smorf)[1], " DE from Ribo-seq")
paste0(sum(sig_res_riboseq_smorf$log2FoldChange > 0), " up from Ribo-seq")
paste0(sum(sig_res_riboseq_smorf$log2FoldChange < 0), " down from Ribo-seq")

riboseq_volcano_plot <- res_riboseq %>%
  as_tibble(rownames = "geneid") %>%
  drop_na(padj) %>%
  filter(str_detect(geneid, "NR|XR")) %>%
  mutate(diff_riboseq = case_when(
    log2FoldChange >= log2(fold_change_threshold) & padj < pval_threshold & baseMean >= base_mean_threshold & str_detect(geneid, "NR|XR") == "TRUE" ~ "Up",
    log2FoldChange <= -log2(fold_change_threshold) & padj < pval_threshold & baseMean >= base_mean_threshold & str_detect(geneid, "NR|XR") == "TRUE" ~ "Down",
    TRUE ~ "No change"
  )) %>%
  ggplot(aes(x = log2FoldChange, y = -log10(padj), color = diff_riboseq)) +
  geom_point(size = 1) +
  labs(x = "Log2 fold change (TS/NTS)", title = "Ribo-seq", color = "") +
  theme_bw() +
  xlim(-2.5, 2.5) +
  geom_vline(
    xintercept = c(log2(fold_change_threshold), -log2(fold_change_threshold)),
    linetype = c(11, 11), size = c(0.2, 0.2)
  ) +
  geom_hline(yintercept = -log10(0.05), linetype = 11, size = 0.2) +
  scale_color_manual(values = c("#D55E00", "#999999", "#56B4E9"))
```

# Comparison of RNA and RPF changes
```{r}
# agreement between ribo and rna
RNASeq_up <- sig_res_rnaseq_smorf %>%
  filter(log2FoldChange > 0) %>%
  dplyr::select(geneid)

RiboSeq_up <- sig_res_riboseq_smorf %>%
  filter(log2FoldChange > 0) %>%
  dplyr::select(geneid)

RNASeq_down <- sig_res_rnaseq_smorf %>%
  filter(log2FoldChange < 0) %>%
  dplyr::select(geneid)

RiboSeq_down <- sig_res_riboseq_smorf %>%
  filter(log2FoldChange < 0) %>%
  dplyr::select(geneid)

length(intersect(RiboSeq_up$geneid, RNASeq_up$geneid))
length(intersect(RiboSeq_down$geneid, RNASeq_down$geneid))

de_list <- list(
  `RNA\nIncrease` = RNASeq_up$geneid,
  `RPF\nIncrease` = RiboSeq_up$geneid,
  `RNA\nDecrease` = RNASeq_down$geneid,
  `RPF\nDecrease` = RiboSeq_down$geneid
)
```
## Venn diagram
```{r}
de_comparison <- ggvenn(
  de_list,
  fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#CD534CFF"),
  stroke_size = 0.5, set_name_size = 4
) +
  scale_fill_viridis(discrete = T)

de_comparison
```

## Figure S9
```{r}

combined_volcano <- (rnaseq_volcano_plot + guides(color = guide_legend(override.aes = list(size = 3))) + riboseq_volcano_plot + guides(color = guide_legend(override.aes = list(size = 3)))) + plot_layout(guides = "collect") 

volcano_venn <- combined_volcano / de_comparison + plot_layout(heights = unit(c(5, 1), c("cm", "null")))
volcano_venn 
ggsave(plot = volcano_venn, filename = paste(results_dir, "Figure_S9.png", sep = ""), width = 9, height = 9, device = "png", dpi = 700)
```

## Volcano plot
```{r}
aggreement_changes_smorf <- c(
  intersect(RNASeq_up$geneid, RiboSeq_up$geneid),
  intersect(RNASeq_down$geneid, RiboSeq_down$geneid)
)

changing_both_smorf <- sig_res_riboseq_smorf %>%
  filter(geneid %in% aggreement_changes_smorf)

riboseq_cor_smorf_rnaseq_volcano_plot <- res_riboseq %>%
  as_tibble(rownames = "geneid") %>%
  filter(str_detect(geneid, "NR|XR")) %>%
  drop_na(padj) %>%
  mutate(diff_ribo = case_when(
    log2FoldChange >= log2(fold_change_threshold) & padj < pval_threshold & baseMean > base_mean_threshold & geneid %in% aggreement_changes_smorf ~ "Up",
    log2FoldChange <= -log2(fold_change_threshold) & padj < pval_threshold & baseMean > base_mean_threshold & geneid %in% aggreement_changes_smorf ~ "Down",
    TRUE ~ "No change"
  )) %>%
  ggplot(aes(x = log2FoldChange, y = -log10(padj), color = diff_ribo)) +
  geom_point(size = 0.5) +
  labs(x = expression("Log"[2] ~ "RPF FC (TS/NTS)"), color = "") +
  theme_bw() +
  geom_vline(xintercept = c(log2(fold_change_threshold), -log2(fold_change_threshold)), linetype = 11, size = 0.2) +
  geom_hline(yintercept = -log10(pval_threshold), linetype = 11, size = 0.2) +
  xlim(-6.5, 6.5) +
  scale_color_manual(values = c("#999999", "#56B4E9")) +
  guides(color = guide_legend(override.aes = list(size = 2))) +
  theme(legend.position = "bottom")
riboseq_cor_smorf_rnaseq_volcano_plot
ggsave(plot = riboseq_cor_smorf_rnaseq_volcano_plot, filename = paste(results_dir, "Figure_6D.png", sep = ""), width = 3, height = 4, device = "png", dpi = 700)
```
## Boxplots of correlated genes
```{r}
sig_res_riboseq_smorf %>%
  filter(geneid %in% aggreement_changes_smorf) %>%
  arrange(-log2FoldChange)

tcounts <- t(log2((counts(dds_te[aggreement_changes_smorf, ], normalized = TRUE, replaced = FALSE) + .5))) %>%
  base::merge(colData(dds_te), ., by = "row.names")

boxplot_smorf <- bind_rows(
  tcounts[tcounts$condition == "nts" & tcounts$assay == "riboseq", aggreement_changes_smorf],
  tcounts[tcounts$condition == "ts" & tcounts$assay == "riboseq", aggreement_changes_smorf],
  tcounts[tcounts$condition == "nts" & tcounts$assay == "rnaseq", aggreement_changes_smorf],
  tcounts[tcounts$condition == "ts" & tcounts$assay == "rnaseq", aggreement_changes_smorf]
) %>%
  mutate(
    condition = toupper(tcounts$condition),
    seqtype = tcounts$assay
  ) %>%
  pivot_longer(cols = c(-condition, -seqtype)) %>%
  # mutate(name = str_remove(name, "_1")) %>%
  ggplot(aes(x = seqtype, y = value, fill = condition)) +
  geom_boxplot(position = position_dodge2(0.75, preserve = "single")) +
  geom_point(size = 0.5, position = position_dodge(0.75, preserve = "total")) +
  facet_wrap(~name, scales = "free_y", ncol = 4) +
  theme_bw() +
  labs(x = "", y = expression("Log"[2] ~ "TE (TS/NTS)"), fill = "") +
  scale_fill_manual(values = c("#D55E00", "#56B4E9")) +
  theme(strip.text = element_text(face = "italic", size = 7), strip.background = element_blank())

boxplot_smorf
ggsave(plot = boxplot_smorf, filename = paste(results_dir, "Figure_S10.png", sep = ""), width = 10, height = 10, device = "png", dpi = 700)
```
# Differential smORF translation
```{r}
sig_res_te_smorf <- res_te %>%
  as_tibble(rownames = "geneid") %>%
  filter(str_detect(geneid, "NR|XR")) %>%
  filter(geneid %in% smorf_ids$`ORF-RATER name`) %>%
  dplyr::filter(abs(log2FoldChange) > log2(fold_change_threshold) & padj < pval_threshold & baseMean >= base_mean_threshold) %>%
  arrange(-log2FoldChange)
sig_res_te_smorf
```
## TE boxplot
```{r}
goi <- sig_res_te_smorf$geneid

tcounts <- t(log2((counts(dds_te[goi, ], normalized = TRUE, replaced = FALSE) + .5))) %>%
  merge(colData(dds_te), ., by = "row.names")

smorf_te_boxplots <- bind_rows(
  tcounts[tcounts$condition == "nts" & tcounts$assay == "riboseq", goi] /
    tcounts[tcounts$condition == "nts" & tcounts$assay == "rnaseq", goi],
  tcounts[tcounts$condition == "ts" & tcounts$assay == "riboseq", goi] /
    tcounts[tcounts$condition == "ts" & tcounts$assay == "rnaseq", goi]
) %>%
  mutate(condition = toupper(tcounts$condition[1:8])) %>%
  pivot_longer(cols = -condition) %>%
  mutate(name = str_remove(name, "_1")) %>%
  ggplot(aes(x = condition, y = value, fill = condition)) +
  geom_boxplot() +
  geom_point(size = 0.5) +
  facet_wrap(~name, scales = "free_y", ncol = 3) +
  theme_bw() +
  labs(x = "", y = expression("Log"[2] ~ "TE (TS/NTS)"), fill = "") +
  scale_fill_manual(values = c("#D55E00", "#56B4E9")) +
  theme(strip.text = element_text(face = "italic", size = 8), strip.background = element_blank(), axis.title = element_text(size = 6), legend.position = "bottom", axis.title.y = element_text(size = 12))
smorf_te_boxplots
ggsave(smorf_te_boxplots, filename = paste(results_dir, "Figure_6E.png", sep = ""), width = 7, height = 5, device = "png", dpi = 700)
```
## Write gene lists to file
```{r}
fn <- paste(results_dir, "Table S6.xlsx",
  sep = ""
)

suppressMessages(if (file.exists(fn)) {
  file.remove(fn)
})

write_xlsx(list(
  lncORF_DE_RNASeq = sig_res_rnaseq_smorf,
  lncORF_DE_RiboSeq = sig_res_riboseq_smorf,
  lncORF_DTE = sig_res_te_smorf
),
path = fn,
format_headers = TRUE
)
```
