---
title: "Section 2.6 Results"
output:
  html_document:
    df_print: paged
---

# Microproteins are differentially expressed between the exponential and stationary phases of CHO cell culture

## Prepare for analysis
### Load libraries & functions

```{r include=FALSE}
package_list <- c(
  "tidyverse", "proDA", "scales", "pheatmap", 
  "readxl","writexl", "viridis", "cowplot"
)

lapply(package_list, require, character.only = TRUE)
```

### Create results directory
```{r include=FALSE}
results_dir <- "../section2.6/"
if (!dir.exists(results_dir)) {
  dir.create(results_dir, recursive = TRUE)
}
```

### Import the ORFRATER results
```{r}
load("../section2.2/results_2_2.RData")
```


# CHOK1GS growth rate data
```{r warning=FALSE}
growth_rates <- read_excel("../../data/CHOK1GS_growth_rate.xlsx",
  col_types = c(
    "text", "text", "numeric",
    "numeric", "numeric", "numeric",
    "numeric", "numeric", "numeric"
  )
) %>%
  pivot_longer(names_to = "Day", cols = c("Day 1", "Day 2", "Day 3", "Day 4", "Day 5", "Day 6", "Day 7", ))

growth_rate_plot <- growth_rates %>%
  ggplot(aes(x = Day, y = value, fill = Day)) +
  geom_vline(xintercept = c("Day 4", "Day 7"), linetype = 11, size = 0.2) +
  geom_boxplot() +
  scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
  scale_fill_manual(values = c("#999999", "#999999", "#999999", "#8785BA", "#999999", "#999999", "#5AAA46")) +
  theme_bw() +
  theme(legend.position = "none") +
  labs(x = "", y = "Density (cells/ml)")
growth_rate_plot
ggsave(growth_rate_plot, filename = paste0(results_dir,"Figure 6A.png"), device = "png", dpi = 700, width = 4.5, height = 3)
```

## Proteomics
### Load and filter the results
```{r echo=FALSE, message=TRUE, warning=TRUE}
d4d7_full <- read_excel("../../data/CHOK1GS_D4_v_D7.xlsx",) %>%
  filter(!grepl("Confidence|High", Accession)) # filter out peptide level rows

# select "uniprot" proteins > 100aa with 2 or more identifications
d4d7_protein_uniprot_long <- d4d7_full %>%
  filter(!grepl("XR_|XM_", Accession)) %>% # remove proteoforms identified by ORF-RATER
  filter(as.numeric(`# AAs`) >= 100 & as.numeric(`# Peptides`) >= 2)

paste0(dim(d4d7_protein_uniprot_long)[1], " standard uniprot proteins identified")

# select "uniprot" proteins <= 100aa with 1 or more identifications
d4d7_protein_uniprot_short <- d4d7_full %>%
  filter(!grepl("XR_|XM_", Accession)) %>% # remove proteoforms identified by ORF-RATER
  filter(as.numeric(`# AAs`) < 100 & as.numeric(`# Peptides`) >= 1)
paste0(dim(d4d7_protein_uniprot_short)[1], " uniprot proteins < 100aa identified")

# select "orfrater" proteins > 100aa with 2 or more identifications
d4d7_orfrater_long <- d4d7_full %>%
  filter(grepl("XR_|XM_", Accession)) %>%
  filter(as.numeric(`# AAs`) >= 100 & as.numeric(`# Peptides`) >= 2)

paste0(dim(d4d7_orfrater_long)[1], " ORF-RATER identified proteoforms > 100aa")

# select "orfrater" proteins <= 100aa with 1 or more identifications
d4d7_orfrater_short <- d4d7_full %>%
  filter(grepl("XR_|XM_", Accession)) %>%
  filter(as.numeric(`# AAs`) < 100 & as.numeric(`# Peptides`) >= 1)

paste0(dim(d4d7_orfrater_short)[1], " ORR-RATER identified proteoforms  <= 100aa")

sum(dim(d4d7_protein_uniprot_long)[1], dim(d4d7_protein_uniprot_short)[1], dim(d4d7_orfrater_long)[1], dim(d4d7_orfrater_short)[1])


identified_proteoforms_d4d7 <- data.frame(
  experiment = c("D4D7", "D4D7"),
  orf_type = c("< 100 aa", ">= 100 aa"),
  count = c(dim(d4d7_orfrater_short)[1], dim(d4d7_orfrater_long)[1])
)
```


```{r}
comparison_barplot <- bind_rows(identified_proteoforms_d4d7) %>%
  arrange(desc(orf_type)) %>%
  ggplot(aes(x = experiment, y = count, fill = orf_type)) +
  geom_bar(stat = "identity") +
  labs(x = "", y = "# proteins", fill = "Novel ORF class") +
  theme_minimal() +
  scale_y_continuous(expand = c(0, 0), breaks = c(0, 16, 95)) +
  scale_fill_viridis(discrete = T, direction = 1) +
  coord_flip() +
  theme(legend.position = "top", axis.text.y = element_blank())

comparison_barplot
ggsave(comparison_barplot, filename = paste0(results_dir,"Figure 6C.png"), device = "png", dpi = 700, width = 4, height = 1.5)
```
### Merge with ORF-RATER output
```{r}
orfrater_d4d7_ids <- bind_rows(d4d7_orfrater_short)

d4d7_orfrater_ms_merged <- table_s3 %>%
  mutate(Accession = `ORF-RATER name`) %>%
  left_join(orfrater_d4d7_ids, by = "Accession") %>%
  drop_na(Description)


sum(table(d4d7_orfrater_ms_merged$`ORF type`, d4d7_orfrater_ms_merged$`Start codon`))

d4d7_orfrater_ms_merged %>%
  filter(`ORF type` == "New")

table(d4d7_orfrater_ms_merged$`ORF type`)

orftype_codon_data <- data.frame(table(d4d7_orfrater_ms_merged$`ORF type`, d4d7_orfrater_ms_merged$`Start codon`)) %>%
  mutate(Var3 = gsub("T", "U", Var2)) %>%
  group_by(Var3) %>%
  mutate(count_name_occur = Freq) %>%
  group_by(Var1) %>%
  mutate(total = sum(Freq)) %>%
  mutate(codons = factor(Var3, levels = c("UUG", "GUG", "CUG", "AUG")))
```

### Plot by codon and ORF type
```{r}
d4d7_orftype_codon_plot <- orftype_codon_data %>%
  ggplot(aes(x = reorder(Var1, count_name_occur), y = Freq, fill = codons)) +
  geom_bar(stat = "identity") +
  scale_fill_viridis(discrete = T, direction = -1) +
  theme_minimal_vgrid() +
  coord_flip() +
  labs(y = "# Identified ORFs", x = "", fill = "Start codon") +
  theme(legend.position = "top") +
  scale_y_continuous(breaks = pretty_breaks())

d4d7_orftype_codon_plot
ggsave(d4d7_orftype_codon_plot, filename = paste0(results_dir, "Figure 6D.png"), device = "png", dpi = 700, width = 5, height = 4)
```

## Differential expression analysis using proDA
```{r}
# filter proteins by length and number of peptides detected, rename samples
d4d7_de <- d4d7_full %>%
  filter((as.numeric(`# AAs`) >= 100 & as.numeric(`# Peptides`) >= 2) | (as.numeric(`# AAs`) < 100 & as.numeric(`# Peptides`) >= 1)) %>%
  mutate(
    `day4_r1` = as.numeric(`Abundance: F14: Sample`),
    `day4_r2` = as.numeric(`Abundance: F15: Sample`),
    `day4_r3` = as.numeric(`Abundance: F16: Sample`),
    `day4_r4` = as.numeric(`Abundance: F17: Sample`),
    `day7_r1` = as.numeric(`Abundance: F18: Sample`),
    `day7_r2` = as.numeric(`Abundance: F19: Sample`),
    `day7_r3` = as.numeric(`Abundance: F20: Sample`),
    `day7_r4` = as.numeric(`Abundance: F21: Sample`)
  )

# create a matrix to hold the raw abundance values
abundance_matrix_d4d7 <- as.matrix(d4d7_de %>%
  select(c(
    "day4_r1", "day4_r2", "day4_r3", "day4_r4",
    "day7_r1", "day7_r2", "day7_r3", "day7_r4"
  )))

# add row names
rownames(abundance_matrix_d4d7) <- d4d7_de$Accession


# change zero to NA for proDA
abundance_matrix_d4d7[abundance_matrix_d4d7 == 0] <- NA

# transform and normalize
abundance_matrix_d4d7 <- log2(abundance_matrix_d4d7)
normalized_abundance_matrix_d4d7 <- median_normalization(abundance_matrix_d4d7)

# create a design matrix
sample_info_df_d4d7 <- data.frame(
  name = colnames(normalized_abundance_matrix_d4d7),
  stringsAsFactors = FALSE
)

sample_info_df_d4d7$condition <- c("d4", "d4", "d4", "d4", "d7", "d7", "d7", "d7")
sample_info_df_d4d7$replicate <- c(1, 2, 3, 4, 1, 2, 3, 4)

# fit with proDA
fit_d4d7 <- proDA(normalized_abundance_matrix_d4d7,
  design = ~condition,
  col_data = sample_info_df_d4d7, reference_level = "d4"
)
```

## Heatmap

```{r}
# plot heatmap of transformed abundances
d4_d7_heatmap <- pheatmap(dist_approx(fit_d4d7, by_sample = TRUE)$mean,
  color = viridis(100),
  labels_row = c("D4_R1", "D4_R2", "D4_R3", "D4_R4", "D7_R1", "D7_R2", "D7_R3", "D7_R4"),
  labels_col = c("D4_R1", "D4_R2", "D4_R3", "D4_R4", "D7_R1", "D7_R2", "D7_R3", "D7_R4")
)

png(paste0(results_dir,"Figure 6E.png"), units = "in", width = 5, height = 5, res = 700)
d4_d7_heatmap
dev.off()
```


```{r}
d4_v_d7_proDA_result <- test_diff(fit_d4d7, contrast = "conditiond7")

d4d7_de_proteins_volcano <- d4_v_d7_proDA_result %>%
  mutate(Accession = name) %>%
  left_join(d4d7_full, by = "Accession") %>%
  mutate(log10pvalue = -log10(adj_pval)) %>%
  mutate(protein_type = case_when(
    grepl("XR_|XM_", name) & adj_pval < 0.05 & abs(diff) >= log(1.5) & n_obs >= 4 & as.numeric(`# AAs`) < 100 ~ "Microprotein",
    grepl("XR_|XM_", name) & adj_pval < 0.05 & abs(diff) >= log(1.5) & n_obs >= 4 & as.numeric(`# AAs`) >= 100 ~ "Novel proteform",
    !grepl("XR_|XM_", name) & adj_pval < 0.05 & abs(diff) >= log(1.5) & n_obs >= 4 ~ "Uniprot",
    adj_pval >= 0.05 | abs(diff) < log(1.5) | n_obs < 4 ~ "Non DE"
  )) %>%
  mutate(protein_alpha = case_when(
    grepl("XR_|XM_", name) & adj_pval < 0.05 & abs(diff) >= log(1.5) & n_obs >= 4 ~ 1,
    !grepl("XR_|XM_", name) & adj_pval < 0.05 & abs(diff) >= log(1.5) & n_obs >= 4 ~ 0.5,
    adj_pval >= 0.05 | abs(diff) < log(1.5) | n_obs < 4 ~ 0.1
  )) %>%
  mutate(protein_size = case_when(
    grepl("XR_|XM_", name) & adj_pval < 0.05 & abs(diff) >= log(1.5) & n_obs >= 4 ~ 1,
    !grepl("XR_|XM_", name) & adj_pval < 0.05 & abs(diff) >= log(1.5) & n_obs >= 4 ~ 0.1,
    adj_pval >= 0.05 | abs(diff) < log(1.5) | n_obs < 4 ~ 0.1
  )) %>%
  arrange(desc(protein_type))

table(d4d7_de_proteins_volcano$protein_type)

d4d7_volcano_plot <- d4d7_de_proteins_volcano %>%
  # filter(protein_type == "Novel ORF") %>%
  ggplot(aes(x = diff, y = log10pvalue, color = protein_type, label = name)) +
  geom_point(size = d4d7_de_proteins_volcano$protein_size, alpha = d4d7_de_proteins_volcano$protein_alpha) +
  # geom_text_repel(size=3) +
  geom_vline(xintercept = c(log(1.5), -log(1.5)), linetype = 11, size = 0.2) +
  geom_hline(yintercept = c(-log10(0.05)), linetype = 11, size = 0.2) +
  scale_color_manual(values = c("#DC3220", "#999999", "#006CD1", "#E1BE6A")) +
  theme_bw() +
  lims(x = c(-4, 4)) +
  labs(x = "Log protein FC (Day 7/Day 4)", y = "-Log10 (padj)", color = "") +
  guides(color = guide_legend(override.aes = list(size = 1))) +
  theme(legend.position = "right")

d4d7_volcano_plot
ggsave(d4d7_volcano_plot, filename = paste0(results_dir,"Figure 6F.png"), device = "png", dpi = 700, width = 4, height = 3.5)
```

## Table S7
```{r}
# create a new frame to hold the DE proteins and merge with the original proteomics results
d4d7_de_proteins <- d4_v_d7_proDA_result %>%
  filter(adj_pval < 0.05 & abs(diff) >= log(1.5) & n_obs >= 4) %>%
  # filter(grepl("XR_|XM_",name)) %>%
  filter(diff <= -log(1.5)) %>%
  dplyr::rename(
    Accession = name, logFC = diff, t = t_statistic,
    P.Value = pval, adj.P.Val = adj_pval
  ) %>%
  select(Accession, logFC, t, P.Value, adj.P.Val, n_obs)

unprot_identifications <- bind_rows(d4d7_protein_uniprot_long, d4d7_protein_uniprot_short) %>%
  select(-c("Checked", "Master", "Contaminant", "Marked as", "...35", "...36", "...37", "...38", "...39", "...40", "...41")) %>%
  rename(
    `D4_R1` = `Abundance: F14: Sample`,
    `D4_R2` = `Abundance: F15: Sample`,
    `D4_R3` = `Abundance: F16: Sample`,
    `D4_R4` = `Abundance: F17: Sample`,
    `D7_R1` = `Abundance: F18: Sample`,
    `D7_R2` = `Abundance: F19: Sample`,
    `D7_R3` = `Abundance: F20: Sample`,
    `D7_R4` = `Abundance: F21: Sample`,
    `Found in Sample: D4_R1` = `Found in Sample: [S14] F14: Sample`,
    `Found in Sample: D4_R2` = `Found in Sample: [S15] F15: Sample`,
    `Found in Sample: D4_R3` = `Found in Sample: [S16] F16: Sample`,
    `Found in Sample: D4_R4` = `Found in Sample: [S17] F17: Sample`,
    `Found in Sample: D7_R1` = `Found in Sample: [S18] F18: Sample`,
    `Found in Sample: D7_R2` = `Found in Sample: [S19] F19: Sample`,
    `Found in Sample: D7_R3` = `Found in Sample: [S20] F20: Sample`,
    `Found in Sample: D7_R4` = `Found in Sample: [S21] F21: Sample`,
  )

write_xlsx(unprot_identifications, path = paste0(results_dir,"Table S7A.xlsx"))

orfrater_identifications <- bind_rows(d4d7_orfrater_short, d4d7_orfrater_long) %>%
  select(-c("Checked", "Master", "Contaminant", "Marked as", "...35", "...36", "...37", "...38", "...39", "...40", "...41")) %>%
  rename(
    `D4_R1` = `Abundance: F14: Sample`,
    `D4_R2` = `Abundance: F15: Sample`,
    `D4_R3` = `Abundance: F16: Sample`,
    `D4_R4` = `Abundance: F17: Sample`,
    `D7_R1` = `Abundance: F18: Sample`,
    `D7_R2` = `Abundance: F19: Sample`,
    `D7_R3` = `Abundance: F20: Sample`,
    `D7_R4` = `Abundance: F21: Sample`,
    `Found in Sample: D4_R1` = `Found in Sample: [S14] F14: Sample`,
    `Found in Sample: D4_R2` = `Found in Sample: [S15] F15: Sample`,
    `Found in Sample: D4_R3` = `Found in Sample: [S16] F16: Sample`,
    `Found in Sample: D4_R4` = `Found in Sample: [S17] F17: Sample`,
    `Found in Sample: D7_R1` = `Found in Sample: [S18] F18: Sample`,
    `Found in Sample: D7_R2` = `Found in Sample: [S19] F19: Sample`,
    `Found in Sample: D7_R3` = `Found in Sample: [S20] F20: Sample`,
    `Found in Sample: D7_R4` = `Found in Sample: [S21] F21: Sample`,
  ) %>%
  rename(`ORF-RATER name` = Accession) %>%
  left_join(table_s3, by = "ORF-RATER name")
write_xlsx(orfrater_identifications, path = paste0(results_dir,"Table S7B.xlsx"))

annotated_d4_d7_de_table <- read_excel("../../data/CHOK1GS_D4_v_D7.xlsx") %>%
  filter(!grepl("Confidence|High", Accession)) %>%
  left_join(d4d7_de_proteins, by = "Accession") %>%
  drop_na(P.Value) %>%
  filter(n_obs >= 4) %>%
  arrange(logFC) %>%
  select(-c("Checked", "Master", "Contaminant", "Marked as", "...35", "...36", "...37", "...38", "...39", "...40", "...41")) %>%
  rename(
    `D4_R1` = `Abundance: F14: Sample`,
    `D4_R2` = `Abundance: F15: Sample`,
    `D4_R3` = `Abundance: F16: Sample`,
    `D4_R4` = `Abundance: F17: Sample`,
    `D7_R1` = `Abundance: F18: Sample`,
    `D7_R2` = `Abundance: F19: Sample`,
    `D7_R3` = `Abundance: F20: Sample`,
    `D7_R4` = `Abundance: F21: Sample`,
    `Found in Sample: D4_R1` = `Found in Sample: [S14] F14: Sample`,
    `Found in Sample: D4_R2` = `Found in Sample: [S15] F15: Sample`,
    `Found in Sample: D4_R3` = `Found in Sample: [S16] F16: Sample`,
    `Found in Sample: D4_R4` = `Found in Sample: [S17] F17: Sample`,
    `Found in Sample: D7_R1` = `Found in Sample: [S18] F18: Sample`,
    `Found in Sample: D7_R2` = `Found in Sample: [S19] F19: Sample`,
    `Found in Sample: D7_R3` = `Found in Sample: [S20] F20: Sample`,
    `Found in Sample: D7_R4` = `Found in Sample: [S21] F21: Sample`,
  ) %>%
  select(c(Accession, Description, logFC, t, P.Value, adj.P.Val, n_obs, everything()))

write_xlsx(annotated_d4_d7_de_table, path = paste0(results_dir,"Table S7C.xlsx"))
```
